Home: Fix unhandled exception on layout query

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/dotcom-forge/issues/3601
## Proposed Changes

This PR fixes the issue mentioned at p1693373541277569-slack-C04U5A26MJB.

### Why this error occurred?
Inside the `Home` component we use a hook to fetch the home layout ( `useHomeLayoutQuery` ).
If the request fails, we end up having a `layout` property that is `null`, the layout components are trying to load and we try to access the properties of an object that doesn't exist.

## Testing Instructions

### Test 1: Functionality
1. Apply the following diff ( not on this branch ).
```diff --git a/client/data/home/use-home-layout-query.ts b/client/data/home/use-home-layout-query.ts
index f089d41a11..f02d329e7d 100644
--- a/client/data/home/use-home-layout-query.ts
+++ b/client/data/home/use-home-layout-query.ts
@@ -29,6 +29,9 @@ export function fetchHomeLayout(
 	siteId: number | null,
 	query: HomeLayoutQueryParams = {}
 ): Promise< unknown > {
+	return new Promise( ( resolve, reject ) => {
+		reject( new Error( 'Emulating an error....' ) );
+	} );
 	return wpcom.req.get(
 		{
 			path: `/sites/${ siteId }/home/layout`,
```
<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
2. Visit `/home`.
3. Open the dev console and wait for a couple of seconds.
4. Ensure that you see the following error and the page doesn't load:
![emu_error](https://github.com/Automattic/wp-calypso/assets/497103/282c9563-deb5-417b-a08a-b9f17e4007b2)
5. Checkout this PR, keeping the applied diff.
6. Visit `/home`
7. Ensure that the page is not crashing.

### Test 2: Analytics
1. Checkout this PR, keeping the applied diff.
2. Enable debugging to be able to see analytic events.
3. Visit `/home`
4. Ensure that the page is not crashing.
5. Ensure that you see `calypso_customer_home_my_site_view_layout_error` being recorded.


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

kozer committed:
Home: Fix unhandled exception on layout query

danielbachhuber reviewed (commented):
Couple of questions:
1. Does this issue impact other views as well, or just Home?
2. At what point was this issue introduced?

kozer commented:
> Does this issue impact other views as well, or just Home?
`useHomeLayoutQuery` is not used by any other page, and from a quick search, there is not any other "fetching layout" query elsewhere to check if we do the same thing.
> At what point was this issue introduced?
It seems that this was there from the first time this file was created (2021-07-01), but there is also a chance that there is something changed internally on `useQuery` library ( we updated the library on 2023-04-27 and we moved away from the previous API ( passing the query function directly to `useQuery` ), to `queryFn` on 2023-05-24 )

danielbachhuber reviewed (commented):
Couple more questions:
* If `layout` is undefined, what happens to the presentation of the page?
* Are there any other fixes worth considering?

kozer commented:
> If the layout is undefined, what happens to the presentation of the page?
It blows up completely. If you try to reproduce the tests I added, you will see that nothing is displayed.

> Are there any other fixes worth considering?
Not a different solution, but a companion to this one, I'd say. I would add a catch clause and return an empty object in case this endpoint fails. By doing that, we can have 2 guards in place, and avoid breakage if we reuse this hook elsewhere.

kozer committed:
Update: Use error as an extra failsafe guard

kozer commented:
I updated this ticket and used the error as a failsafe. 
Note that we could use `onError` as we do in other places to set a state variable or something as well, but `react-query` docs state that `onError`, `onSuccess` and `onSettled` functions [are deprecated](https://tanstack.com/query/v4/docs/react/reference/useQuery).

kozer committed:
Merge branch 'trunk' into fix/home_layout_unhandled_exception

danielbachhuber commented on client/my-sites/customer-home/main.jsx:
```diff
@@ -178,21 +178,20 @@ const Home = ( {
-				<div className="customer-home__loading-placeholder"></div>
-			) : (
+			{ isLoading && <div className="customer-home__loading-placeholder"></div> }
+			{ ! isLoading && ! homeLayoutError ? (
```
```suggestion
			{ ! isLoading && layout && ! homeLayoutError ? (
```

Should we check for `layout` here, instead of making it optional below? Is there some scenario where `! homeLayoutError` but `layout` is still empty?

kozer replied:
The problem is not if it's empty, but if it's `undefined` or `null`. However, as I don't trust js, I totally agree with your suggestion!

danielbachhuber commented on client/my-sites/customer-home/main.jsx:
```diff
@@ -71,7 +71,7 @@ const Home = ( {
 	const translate = useTranslate();
 
-	const { data: layout, isLoading } = useHomeLayoutQuery( siteId );
+	const { data: layout, isLoading, error: homeLayoutError } = useHomeLayoutQuery( siteId );
```
Should we track how often these errors occur? It seems like we should want to know how often the user is experiencing this broken state, instead of ignoring it entirely.

kozer replied:
You are right! On it!

danielbachhuber reviewed (commented):
Thanks for the detailed testing instructions! I confirm that this PR produces the following screen instead of a blank page:

<img width="1230" alt="CleanShot 2023-09-07 at 04 58 06@2x" src="https://github.com/Automattic/wp-calypso/assets/36432/85da92d4-4483-4aa0-99c3-6b25aaab44e8">

Couple more comments to work through.

kozer committed:
Merge branch 'trunk' into fix/home_layout_unhandled_exception

kozer committed:
Update: Add event when there is a layout error

kozer committed:
Update: update event property

kozer commented:
@danielbachhuber , I created a new pr https://github.com/Automattic/tracks-events-registration/pull/1825/files for the new event. However, [in the tracks view](https://mc.a8c.com/tracks/events/event.php?eventname=calypso_customer_home_my_site_view_layout_error), I can't see its properties. Is it because I haven't merged it yet?
![register_event](https://github.com/Automattic/wp-calypso/assets/497103/0dff6609-7676-4e2a-a61d-8b9ebc65729c)

kozer commented:
Events registered successfully, so we can merge it I guess @danielbachhuber

kozer committed:
Merge branch 'trunk' into fix/home_layout_unhandled_exception
